shotover/src/server.rs

```diff
@@ -255,6 +263,138 @@ pub struct Handler<C: CodecBuilder> {
+                                                                          // close message to the
+                                                                          // client, protocol error
+                                        warn!("failed to decode message: {:?}", err);
+                                        return;
+                                    }
+                                    Err(CodecReadError::Io(_err)) => {
+                                        unreachable!("CodecReadError::Io should not occur because we are reading from a newly created BytesMut")
+                                    }
+                                }
+                            }
+                            Ok(_ws_message) => {
+                                // TODO we need to tell the client about a protocol error
+                                todo!();
+                            }
+                            Err(err) => {
+                                // TODO
+                                error!("{err}");
+                                return;
+                            }
+                        }
+                    } else {
+                        return;
+                    }
+                }
+                _ = in_tx.closed() => {
+                    // main task has shutdown down, this task is no longer needed
+                    return;
+                }
+            }
+        }
+    }
+    .in_current_span(),
+    );
+
+    // write task
+    tokio::spawn(
+        async move {
+            loop {
+                if let Some(message) = out_rx.recv().await {
+                    let mut bytes = BytesMut::new();
+                    match encoder.encode(message, &mut bytes) {
+                        Err(err) => {
+                            error!("failed to encode message destined for client: {err:?}")
+                        }
+                        Ok(_) => {
+                            let message = WsMessage::binary(bytes);
+                            match writer.send(message).await {
+                                Ok(_) => {}
+                                Err(err) => {
+                                    // TODO
+                                    error!("{err}");
+                                    return;
+                                }
+                            }
+                        }
+                    }
+                } else {
+                    // Main task has ended.
+                    // First flush out any remaining messages.
+                    // Then end the task thus closing the connection by dropping the write half
+                    while let Ok(message) = out_rx.try_recv() {
+                        let mut bytes = BytesMut::new();
+                        match encoder.encode(message, &mut bytes) {
+                            Err(err) => {
+                                error!("failed to encode message destined for client: {err:?}")
```

Comment:
this should terminate the connection, if we miss a response then the connection is no longer usable.